### PR TITLE
Fixes #36296 - fix rendering/setting grub password

### PR DESCRIPTION
**Pull request by bmagistro**
Root cause isn't fully understood but this change does resolve the issue being described/experienced.

Fixes: 36296
Relates: 32394 / #8461

### CODE REVIEW

File: app/services/foreman/renderer/scope/macros/host_template.rb [hunk 1]
```diff
@@ -124,7 +124,8 @@ def root_pass
           end
           def grub_pass
             return '' unless host_param_true?('encrypt_grub')
```
Comment (ofedoren) [inline app/services/foreman/renderer/scope/macros/host_template.rb:126]:
```suggestion
```

File: app/services/foreman/renderer/scope/macros/host_template.rb [hunk 2]
```diff
@@ -124,7 +124,8 @@ def root_pass
...
-            host.grub_pass.start_with?('$1$') ? "--md5pass=#{host.grub_pass}" : "--iscrypted --password=#{host.grub_pass}"
+            param_grub_pass = host_param('grub_pass')
+            param_grub_pass.start_with?('$1$') ? "--md5pass=#{param_grub_pass}" : "--iscrypted --password=#{param_grub_pass}"
```
Comment (ofedoren) [inline app/services/foreman/renderer/scope/macros/host_template.rb:128]:
```suggestion
            grub_password = host_param('grub_pass') || host.grub_pass
            return '' unless host_params_true?('encrypt_grub') && grub_password

            grub_password.start_with?('$1$') ? "--md5pass=#{param_grub_pass}" : "--iscrypted --password=#{param_grub_pass}"
```

**Review by ofedoren — CHANGES_REQUESTED** [review]
Thanks, @bmagistro, but the change is not quite right.

Macro `grub_pass` actually uses `root_pass` of a host. It should just work for managed hosts, since the password is required. The password gets encrypted and set as password for both host and grub before the host gets saved.

In case you don't set a password for a host, this macro will simply fail.

In case you want to have grub password be different, you don't have to use this macro, but modify the template directly:
```erb
bootloader [opts] <%= grub_pass %>
```
change to
```erb
bootloader [opts] --iscrypted --password=<%= host_param('grub_pass') %>
```

If you want to be able to set a different grub password for hosts, it would require a bit more changes (model/API/form), but just for now we could limit the scope for the macro and host params only. Please, see suggested changes inline:

**Comment by bmagistro** [comment]
@ofedoren thanks for the feedback.  I see this as two things now.

1) the original issue of `encrypt_grub` not working as expected (not actually addressed by this change).  We do have the root password set (should be inherited on the host) so based on the description, adding the host param to enable this _should_ have been sufficient.  Need to track down the root cause and fix that (more details in the original bbug report)

Looking at the enc_preview, I do see root_pw in there but do not see anything related to grub so that might be a line of inquiry (still figuring out how this should be working)

2) i am very much of the opinion the user should not need to clone a template just to change a password (or a number of other simple things -- ref #9340; stalled as I don't think I had a clear answer on the desired behavior of find_snippet) . there exists provisions to override the root password at various levels.  i believe similar provisions (host parameter?) should be added/exposed to allow the grub password to be different than the root password if desired.  I believe this is very cleanly solved by your suggestions in the short term.  however I think levering this would just mask the issue that caused us to start down this path.

----

so  if you are good with it, figure out what is happening and fix item 1 then as a second commit to keep things better organized implement the or logic for the password described in 2 incorporating your suggestions.

----

edit: follow on to 2 is that I would also like to be able to enable this post install (have a manifest developed and working great to do so) but would ideally only need to track the password in a single field (root_pw, host_param, etc).  this lets us set that value and use it both when the host is (re)built or when we need to enable it post install without rebuilding the host

follow on to 1 is that it renders correctly in template preview but errors out when doing the actual render for an install

**Comment by bmagistro** [comment]
declining this pr as it does not address the underlying issue that is being experienced.  can continue discussions/troubleshooting here.  Will redraft some of this (host param aspect) as a separate enhancement

**Comment by ofedoren** [comment]
Sorry, @bmagistro, I've missed one thing before submitting the first review.

I guess we are talking about two things at once here. The first one is an actual bug (root_pass being ignored for grub_pass setting). The second one is a feature you'd like to see (allowing grub_pass to be different from the root_pass). We can use your PR to address both issues (although, the second issue will be addressed with quite a limit).

1. It appears that https://github.com/theforeman/foreman/commit/60d53c2621b3537e61e9082428ebb585f02fd3bc might broke the logic in setting grub_pass from root_pass as the default behaviour. The weird thing is that the bug was noticed just now. It should be fixed by changing:
https://github.com/theforeman/foreman/blob/develop/app/models/concerns/host_common.rb#L156
to
```ruby
self.grub_pass = crypt_pass(self[:grub_pass].presence || self[:root_pass], :grub)
```
This should solve the initial bug when `root_pass` is ignored leading to the error described in your issue.

2. With your PR changing the macro to use `host_param` as grub_pass we can address the second issue described above, but with limitations:
  - The major one is that, even though it will be applied to provisioned host, the grub_pass from `host_param` wouldn't be saved in the database for the provisioned host.
  - There would be no other way to change grub_pass for a host (no API, no host creation form support)

**Comment by bmagistro** [comment]
Circling back to this, that doesn't seem to fix the base issue.  Going to do some more investigating along those lines and hopefully get something out this weekend to address the bug.  I've tried dropping the grub pass out completely so it only has the root pass but still seeing the starts_with error.

edit: confirmed it is when using host group/settings[root] to get the value.  If i set it on the host itself the templates generate however the password is not in a compatible format for grub. ($5$... vs pbkdf)

**Pull request closed without merging**